File: Template/DP/D&C.py
from itertools import *
from typing import *

"""
決策單調性優化DP

Problems:
- 3826. Minimum Partition Score
- 3500. Minimum Cost to Divide Array Into Subarrays (Python TLE, C++ AC)
"""

class Solution:
    def minPartitionScore(self, nums: List[int], k: int) -> int:
        n = len(nums)
        s = list(accumulate(nums, initial=0))

        def cost(l: int, r: int) -> int:  # cost of [l, r)
            v = s[r]  - s[l]
            return v * (v + 1) // 2
        
        f = [float('inf')] * (n + 1)
        nf = [float('inf')] * (n + 1)
        f[0] = 0

        def solve(l: int, r: int, opt_l: int, opt_r: int):
            if l > r:
                return
            mid = (l + r) // 2

            best_val = float('inf')
            opt = -1
            for p in range(opt_l, min(opt_r, mid - 1) + 1):
                val = f[p] + cost(p, mid)
                if val < best_val:
                    best_val = val
                    opt = p
            nf[mid] = best_val

            solve(l, mid - 1, opt_l, opt)
            solve(mid + 1, r, opt, opt_r)
        
        for j in range(1, k + 1):
            solve(j, n, j - 1, n - 1)
            f = nf[:]
        return f[n]


sol = Solution()
print(sol.minPartitionScore([5,1,2,1], 2))  # 25
print(sol.minPartitionScore([1,2,3,4], 1))  # 55
print(sol.minPartitionScore([1,1,1], 3))  # 3

File: Template/DP/D&C.cpp
#include <bits/stdc++.h>
using namespace std;

/*
決策單調性優化DP

Problems:
- 3826. Minimum Partition Score
- 3500. Minimum Cost to Divide Array Into Subarrays (Python TLE, C++ AC)
*/

class Solution {
public:
    long long minPartitionScore(vector<int>& nums, int k) {
        int n = nums.size();

        vector<long long> s(n + 1);
        for (int i = 0; i < n; i++) s[i + 1] = s[i] + nums[i];

        auto w = [&](int l, int r) -> long long {
            long long v = s[r] - s[l];
            return v * (v + 1) / 2;
        };

        vector<long long> f(n + 1, LLONG_MAX / 2), nf(n + 1, LLONG_MAX / 2);
        f[0] = 0;

        auto dfs = [&](this auto&& dfs, int l, int r, int opt_l,
                       int opt_r) -> void {
            if (l > r) return;
            int mid = l + (r - l) / 2;

            long long best_val = LLONG_MAX / 2;
            int opt = -1;
            for (int p = opt_l; p <= min(opt_r, mid - 1); p++) {
                long long v = f[p] + w(p, mid);
                if (v < best_val) {
                    best_val = v;
                    opt = p;
                }
            }

            nf[mid] = best_val;
            dfs(l, mid - 1, opt_l, opt);
            dfs(mid + 1, r, opt, opt_r);
            return;
        };

        for (int j = 1; j <= k; j++) {
            dfs(j, n, j - 1, n - 1);
            swap(f, nf);
        }
        return f[n];
    }
};

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr);

    Solution sol;
    vector<int> nums;
    int k;

    nums = {5,1,2,1};
    cout << sol.minPartitionScore(nums, 2) << endl;  // 25

    nums = {1,2,3,4};
    cout << sol.minPartitionScore(nums, 1) << endl;  // 55

    nums = {1,1,1};
    cout << sol.minPartitionScore(nums, 3) << endl;  // 3
    return 0;
}